src/Utils.php

```diff
@@ -64,4 +62,24 @@ public static function idnUriConvert(UriInterface $uri, $options = 0)
 
         return $uri;
     }
+
+    /**
+     * @param string $domain
+     * @param int    $options
+     * @param array  $info
+     *
+     * @return string|false
+     */
+    private static function idnToAsci($domain, $options, &$info = [])
+    {
+        if (\preg_match('%^[ -~]+$%', $domain) === 1) {
+            return $domain;
+        }
+
+        if (\extension_loaded('intl') && defined('INTL_IDNA_VARIANT_UTS46')) {
+            return \idn_to_ascii($domain, $options, INTL_IDNA_VARIANT_UTS46, $info);
+        }
+
+        return Idn::idn_to_ascii($domain, $options, Idn::INTL_IDNA_VARIANT_UTS46, $info);
```

Comment:
Can this skip using the polyfill class directly? It's marked internal and it shouldn't be used.
Of not, why did you do it this way?

Reply:
Because we need to be able to call the polyfill even if the original extension is installed to work around the extension using a too old version of the C library. Symfony doesn't have a way for us to do this, other to call code that is formally internal.

Reply:
And this commit (included in polyfill-intl-idn 1.17 release) totally stops Guzzle doing anything other than calling the class method directly now..
https://github.com/symfony/polyfill-intl-idn/commit/3bff59ea7047e925be6b7f2059d60af31bb46d6a

Reply:
...which to me suggests it should in fact be polyfill-intl-idn which should be verifying whether the ICU version available on the host machine is usable, or if the polyfill should be used instead, and it not be left to Guzzle to be look into this and decide for itself (therefore 'breaking' Symfony's rules on using code marked internal directly) .. ?

Reply:
@jonnott not possible. Userland implementations cannot replace functions defined by PHP.

Reply:
Oh well, looks like this PR is the best solution we have then. Merge away @mtdowling & co. ;)

Reply:
@jonnott this commit in the polyfill fixes detecting wrongly that `INTL_IDNA_VARIANT_UTS46` is supported by `idn_to_ascii`. It does not change the fact that `idn_to_ascii` is defined by the extension when defined (and so the polyfill could not define it).

Reply:
> Because we need to be able to call the polyfill even if the original extension is installed to work around the extension using a too old version of the C library. Symfony doesn't have a way for us to do this, other to call code that is formally internal.

I wonder if it would be best to remove the polyfill. Library users can that see a deprecation can set `idn_conversion` to false and use their own custom solutions with a middleware or something like that.

Reply:
That's not always possible.  For me, the trigger on the deprecation notice came from a third party library that uses yet another library which is instantiating its own `Guzzle\Client` instance without any sane way for me to reach it to add middleware or set the `idn_conversion` configuration.